### PR TITLE
Add links to the rust documentation for all MIR structs

### PR DESCRIPTION
**Pull request by bollu**
It's nice to have the links as it aids looking up the basic MIR types. I keep returning to this page for a quick reference, and get annoyed each time that I need to then search in the crate :) This should fix my annoyance.

### CODE REVIEW

File: src/mir/index.md [hunk 1]
```diff
@@ -220,27 +220,27 @@ The main MIR data type is `Mir`. It contains the data for a single
...
-- **Basic blocks**: The basic blocks are stored in the field
+- **[Basic blocks]**: The basic blocks are stored in the field
   `basic_blocks`; this is a vector of `BasicBlockData`
   structures. Nobody ever references a basic block directly: instead,
   we pass around `BasicBlock` values, which are
   [newtype'd] indices into this vector.
-- **Statements** are represented by the type `Statement`.
-- **Terminators** are represented by the `Terminator`.
-- **Locals** are represented by a [newtype'd] index type `Local`. The
+- **[Statements]** are represented by the type `Statement`.
```
Comment (LeSeulArtichaut) [inline src/mir/index.md:228]: I think those links should be on the type name itself
```suggestion
- **Statements** are represented by the type [`Statement`].

**Comment by bollu** [comment]
@LeSeulArtichaut fixed! I also added some of the references I had missed the first time around, such as that to `BasicBlockData`

**Review by LeSeulArtichaut — COMMENTED** [review]
Thanks! If I can add just a small other nit, we usually have links like:

```markdown
[`BasicBlock`]

[`BasicBlock`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.BasicBlock.html
```
instead of the quite redundant
```markdown
[`BasicBlock`][BasicBlock]

[BasicBlock]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.BasicBlock.html
```

I'm willing to merge this as-is, but if you could change this it would be great!

File: src/mir/index.md [hunk 2]
```diff
@@ -221,26 +221,28 @@ function (along with sub-instances of Mir for "promoted constants",
...
+  [`basic_blocks`][basicblocks] of the struct [`Body`][Body]; this is a vector 
+  of [`BasicBlockData`][BasicBlockData] structures. Nobody ever references a
+  basic block directly: instead, we pass around [`BasicBlock`][BasicBlock]
+  values, which are [newtype'd] indices into this vector.
+- **Statements** are represented by the type [`Statement`][Statement].
 - **Terminators** are represented by the `Terminator`.
-- **Locals** are represented by a [newtype'd] index type `Local`. The
-  data for a local variable is found in the `Mir` (the `local_decls`
+- **Locals** are represented by a [newtype'd] index type [`Local`][Local]. 
+  The data for a local variable is found in the `Mir` (the `local_decls`
```
Comment (LeSeulArtichaut) [inline src/mir/index.md:231]: Interestingly, this is outdated, `Mir` was renamed to `Body`... two years ago??? 😅

Nothing to do with this PR though

Comment (bollu) [reply src/mir/index.md:231]: @LeSeulArtichaut I'm happy to send a second PR cleaning it up :)

File: src/mir/index.md [hunk 3]
```diff
@@ -221,26 +221,28 @@ function (along with sub-instances of Mir for "promoted constants",
...
   the special "local" representing the return value.
-- **Places** are identified by the enum `Place`. There are a few variants:
+- **Places** are identified by the enum [`Place`]. There are a few 
+  variants:
   - Local variables like `_1`
   - Static variables `FOO`
   - **Projections**, which are fields or other things that "project
-    out" from a base place. So e.g. the place `_1.f` is a projection,
+    out" from a base place. These are represented by the type 
+    [`Projection`][Projection].  So e.g. the place `_1.f` is a projection,
```
Comment (LeSeulArtichaut) [inline src/mir/index.md:240]: I think you missed one here
```suggestion
    [`Projection`].  So e.g. the place `_1.f` is a projection,
```

File: src/mir/index.md [hunk 4]
```diff
@@ -256,3 +258,12 @@ but [you can read about those below](#promoted)).
...
 [newtype'd]: ../appendix/glossary.html#newtype
+[basicblocks](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.basic_blocks)
+[BasicBlock]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.BasicBlock.html
+[BasicBlockData]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.BasicBlockData.html
+[Statement]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Statement.html
+[Terminator]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/terminator/struct.Terminator.html
+[Local]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Local.html
+[Place]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Place.html
+[Rvalue]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html
+[Operand]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Operand.html
```
Comment (LeSeulArtichaut) [inline src/mir/index.md:269]: Unfortunately you'll also have to change them here
```suggestion
[`BasicBlock`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.BasicBlock.html
[`BasicBlockData`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.BasicBlockData.html
[`Statement`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Statement.html
[`Terminator`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/terminator/struct.Terminator.html
[`Local`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Local.html
[`Place`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Place.html
[`Rvalue`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html
[`Operand`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Operand.html
```

File: src/mir/index.md [hunk 5]
```diff
@@ -221,26 +221,28 @@ function (along with sub-instances of Mir for "promoted constants",
 but [you can read about those below](#promoted)).
 
 - **Basic blocks**: The basic blocks are stored in the field
-  `basic_blocks`; this is a vector of `BasicBlockData`
-  structures. Nobody ever references a basic block directly: instead,
-  we pass around `BasicBlock` values, which are
-  [newtype'd] indices into this vector.
-- **Statements** are represented by the type `Statement`.
+  [`basic_blocks`][basicblocks] of the struct [`Body`][Body]; this is a vector
```
[the remark this answers is not in the document]
Comment (bollu) [reply src/mir/index.md]: @LeSeulArtichaut done and rebased.

**Review by LeSeulArtichaut — COMMENTED** [review]
(Feel free to use GitHub's "commit suggestion" feature)

**Comment by bollu** [comment]
thanks for the suggestions, added them in :) Do you want me to rebase and squash history?

**Comment by LeSeulArtichaut** [comment]
No, you don't have to worry about the Git history here, we always squash-merge PRs 🙂

**Review by LeSeulArtichaut — APPROVED** [review]
Thanks ❤️

**Comment by jyn514** [comment]
> No, you don't have to worry about the Git history here, we always squash-merge PRs 🙂

We do? 👀

**Comment by bollu** [comment]
OK, I'm going to send a follow up PR soon cleaning up the text (removing stale references, adding references to the constants mentioned in the code)

**Comment by LeSeulArtichaut** [comment]
> > No, you don't have to worry about the Git history here, we always squash-merge PRs 🙂
> 
> 
> 
> We do? 👀 

I mean, in the `rustc-dev-guide`, not in the main `rust` repo

**Comment by bollu** [comment]
Bizarre, I just built the version of the book to see locally --- the links are broken!

Broken links of rust sources

What's the correct syntax I ought to follow? Terribly sorry for the confusion, I ought to have built locally and checked the output.

File: src/mir/index.md [hunk 6]
```diff
@@ -256,3 +258,12 @@ but [you can read about those below](#promoted)).
 [mir]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/index.html
 [mirmanip]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/index.html
 [newtype'd]: ../appendix/glossary.html#newtype
+[basicblocks](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.basic_blocks)
```
Comment (LeSeulArtichaut) [inline src/mir/index.md:261]: This should be:
```markdown
[basicblocks]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.basic_blocks
```

Comment (bollu) [reply src/mir/index.md:261]: I fixed this at https://github.com/rust-lang/rustc-dev-guide/pull/984, and reflowed the page.

**Comment by LeSeulArtichaut** [comment]
> Terribly sorry for the confusion, I ought to have built locally and checked the output.

Don't be sorry, I don't think this kind of changes deserves building the book locally. Also I should have caught this myself during review.

**Pull request merged**